hnn_core/network.py

```diff
@@ -271,103 +718,73 @@ def _instantiate_feeds(self, n_trials=1):
-                                feed_type=src_type,
-                                target_cell_type=target_cell_type,
-                                params=p_unique[src_type],
-                                gid=gid)
-                        )
-                feed_times[src_type].append(event_times)
-
-        self.feed_times = feed_times
-
-        # Add tonic inputs from param files
-        # do not overwrite tonic_inputs when called from dipole.py
-        if 'tonic' not in self.feed_times:
-            self.feed_times['tonic'] = dict()
-        for cell_type in ['L2Basket', 'L5Basket', 'L2Pyr', 'L5Pyr']:
-            is_tonic_present = [f'Itonic_{p}_{cell_type}_soma' in self.params
-                                for p in ['A', 't0', 'T']]
-            if any(is_tonic_present):
-                if not all(is_tonic_present):
-                    raise ValueError(f'Tonic input must have the amplitude, '
-                                     f'start time and end time specified. One '
-                                     f'or more parameter may be missing for '
-                                     f'cell type {cell_type}')
-                self.add_tonic_input(
-                    cell_type=cell_type,
-                    amplitude=self.params[f'Itonic_A_{cell_type}_soma'],
-                    t0=self.params[f'Itonic_t0_{cell_type}_soma'],
-                    T=self.params[f'Itonic_T_{cell_type}_soma']
-                )
-
-    def add_tonic_input(self, cell_type, amplitude, t0, T):
-        """Create tonic feed for a given cell type.
+                # 'events': list (trials) of list (cells) of list (events)
+                self.external_drives[
+                    drive['name']]['events'].append(event_times)
+
+    def add_tonic_bias(self, *, cell_type=None, amplitude=None,
+                       t0=None, T=None):
+        """Attach parameters of tonic biasing input for a given cell type.
 
         Parameters
         ----------
         cell_type : str
             The cell type whose cells will get the tonic input.
-            Valid inputs include 'L2Basket', 'L5Basket', 'L2Pyr', 'L5Pyr'.
+            Valid inputs are those in `net.cellname_list`.
         amplitude : float
             The amplitude of the input.
         t0 : float
-            The start time of tonic input (in ms).
+            The start time of tonic input (in ms). Default: 0 (beginning of
+            simulation).
         T : float
-            The end time of tonic input (in ms).
+            The end time of tonic input (in ms). Default: end of simulation.
         """
+        if (cell_type is None or amplitude is None):
+            raise ValueError('cell_type and amplitude must be defined'
+                             f', got {cell_type}, {amplitude}')
+        if 'tonic' not in self.external_biases:
+            self.external_biases['tonic'] = dict()
+        if cell_type in self.external_biases['tonic']:
```

Comment:
Hmm so does this still get triggered when you copy the network? Or does `_reset_drives()` take care of that. If not, it may be nice to overwrite specific drives after copying the Network when writing code for parameter sweeps.

Reply:
See my comment to @rythorpe below. The only thing `_reset` are previous simulation results, specifically: `vsoma/isoma` and the `'events'` values (spike times) of drives. It does _not_ remove drives or biases.

I _think_ this is what you want when doing a parameter sweep, right?

Reply:
So if a user first creates a tonic input that targets L2/3 basket cells by calling `net.add_tonic_bias()`, and then post-hoc decides to change the strength of the injected current for this input, how are they supposed to do this? If I understand things correctly, they can't call `net.add_tonic_bias()` again because the L2/3 basket cell type already exists in `net.external_biases['tonic']` right? Alternatively, I guess the user could directly modify `net.external_biases['tonic']`, but if this is the case it should say so in the documentation explicitly (maybe it already does and I just missed it?).

Reply:
If this is a documentation issue rather than a bug, let's definitely save this for a later PR. I'm just unclear at the moment how tonic biases get modified (either in an original `Network` instance or a copied instance).

Reply:
I think the current implementation doesn't explicitly have methods for _ad hoc_ modification, though it's certainly possible by modifying the dicts directly, as you point out (and as is done in the `evoked` example). But both are easy to add, since both drives and biases don't really "materialise" until they're baked into the simulation in `NetworkBuilder`.

Reply:
> ow are they supposed to do this?

To answer your question: either modify the dict, or create a new network and add the correct bias.

Reply:
I think the modification thing is a separate discussion to have :) You can always change one value in your script and re-run those lines ... so not even sure if it's a strong use case

Reply:
What's weird though is that one of the examples shows how to modify the drives ... so it suggests that any other modifications should work. I'm almost inclined to remove the syncing example until we can properly think about how to modify the drives on an existing network.

Reply:
Okay, I'm cool with leaving this to be discussed in more detail later.